### PR TITLE
Implemented Datastorage with overkill

### PR DESCRIPTION
**Pull request by JarnoWesthof**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -83,3 +84,90 @@ deathLinkService.OnDeathLinkReceived += (deathLinkObject) => {
 // ... On death:
 deathLinkService.SendDeathLink(new DeathLink("Ijwu", "Died to exposure."));
 ```
+
+## DataStorage
+
+DataStorage support is included in the library. you may save values on the archipelago server in order to share there across other players or to simply keep track of values outside of your game's state
```
Comment (Ijwu) [inline README.md:90]:
```suggestion
DataStorage support is included in the library. You may save values on the archipelago server in order to share them across other player's sessions or to simply keep track of values outside of your game's state.
```

File: README.md [hunk 2]
```diff
@@ -83,3 +84,90 @@ deathLinkService.OnDeathLinkReceived += (deathLinkObject) => {
...
+
+The DataStorage provides an interface based on keys and thier scope. by assinging a value to a key, that value is stored on the server, and by reading from a key a value is retrieved from the server.
```
Comment (Ijwu) [inline README.md:92]:
```suggestion
The DataStorage provides an interface based on keys and their scope. By assigning a value to a key, that value is stored on the server and by reading from a key a value is retrieved from the server. 
```

File: README.md [hunk 3]
```diff
@@ -83,3 +84,90 @@ deathLinkService.OnDeathLinkReceived += (deathLinkObject) => {
 // ... On death:
 deathLinkService.SendDeathLink(new DeathLink("Ijwu", "Died to exposure."));
 ```
+
+## DataStorage
+
+DataStorage support is included in the library. you may save values on the archipelago server in order to share there across other players or to simply keep track of values outside of your game's state
+
+The DataStorage provides an interface based on keys and thier scope. by assinging a value to a key, that value is stored on the server, and by reading from a key a value is retrieved from the server. 
+The DataStorage also provides methods to retrieve the value of key asynchronously.
```
Comment (Ijwu) [inline README.md:93]:
```suggestion
The DataStorage also provides methods to retrieve the value of a key asynchronously. 
```

File: README.md [hunk 4]
```diff
@@ -83,3 +84,90 @@ deathLinkService.OnDeathLinkReceived += (deathLinkObject) => {
...
 deathLinkService.SendDeathLink(new DeathLink("Ijwu", "Died to exposure."));
 ```
+
+## DataStorage
+
+DataStorage support is included in the library. you may save values on the archipelago server in order to share there across other players or to simply keep track of values outside of your game's state
+
+The DataStorage provides an interface based on keys and thier scope. by assinging a value to a key, that value is stored on the server, and by reading from a key a value is retrieved from the server. 
+The DataStorage also provides methods to retrieve the value of key asynchronously. 
+If your intrested in keeping track of when a value of a certian key is changed you can use the `OnValueChanged` handler to register a callback for when the value gets updated.
```
Comment (Ijwu) [inline README.md:94]:
```suggestion
If you're interested in keeping track of when a value of a certain key is changed you can use the `OnValueChanged` handler to register a callback for when the value gets updated.
```

File: README.md [hunk 5]
```diff
@@ -83,3 +84,90 @@ deathLinkService.OnDeathLinkReceived += (deathLinkObject) => {
...
+* `%`, Gets remainder after dividing left value by right value
+* `^`, Multiply left value by the power of the right value
+* `>>`, Override left with right value. if right value is lower
+* `<<`, Override left with right value. if right value is bigger
+
+Bitwise operations are supported using the following opperations:
+* `+ Bitwise.Xor(x)`, apply logical exclusive OR to the right value using value x
+* `+ Bitwise.Or(x)`, apply logical OR to the right value using value x
+* `+ Bitwise.And(x)`, apply logical AND to the right value using value x
+* `+ Bitwise.LeftShift(x)`, binairy shift left to the left value by x
```
Comment (Ijwu) [inline README.md:116]:
```suggestion
* `+ Bitwise.LeftShift(x)`, binary shift the value to the left by x
```

File: README.md [hunk 6]
```diff
@@ -83,3 +84,90 @@ deathLinkService.OnDeathLinkReceived += (deathLinkObject) => {
...
+* `+ Bitwise.RightShift(x)`, binairy shift left to the right value by x
```
Comment (Ijwu) [inline README.md:117]:
```suggestion
* `+ Bitwise.RightShift(x)`, binary shift the value to the right by x
```

File: README.md [hunk 7]
```diff
@@ -83,3 +84,90 @@ deathLinkService.OnDeathLinkReceived += (deathLinkObject) => {
...
+
+Operation specific callbacks are supported, these get called only once with the results of the current operation:
+* `+ Callback.Add((oldValue, newValue) => {});`, calls this method after your operation or chain of operations are proccesed
+
+mathematical operations, bitwise operations and callbacks can be chained, given the extended syntax with `()` around each operation
```
Comment (Ijwu) [inline README.md:122]:
```suggestion
Mathematical operations, bitwise operations and callbacks can be chained, given the extended syntax with `()` around each operation.
```

File: README.md [hunk 8]
```diff
@@ -83,3 +84,90 @@ deathLinkService.OnDeathLinkReceived += (deathLinkObject) => {
...
+
+examples:
```
Comment (Ijwu) [inline README.md:124]:
```suggestion
Examples:
```

File: README.md [hunk 9]
```diff
@@ -83,3 +84,90 @@ deathLinkService.OnDeathLinkReceived += (deathLinkObject) => {
...
+```csharp
+var session = ArchipelagoSessionFactory.CreateSession("localhost", 38281);
+session.TryConnectAndLogin("Risk of Rain 2", "Ijwu", new Version(2,1,0));
+
+//Initializing
+session.DataStorage.Initialize("B", 20); //Set initial value for B in global scope if it has no value assigned yet
+
+//Storing/Updating
+session.DataStorage[Scope.Slot, "SetPersonal"] = 20; //Set `SetPersonal` to 20, in scope of the current connected user\slot
+session.DataStorage[Scope.Global, "SetGlobal"] = 30; //Set `SetGlobal` to 30, in gloval scope shared amoung all players (the default scope is global)
```
Comment (Ijwu) [inline README.md:134]:
```suggestion
session.DataStorage[Scope.Global, "SetGlobal"] = 30; //Set `SetGlobal` to 30, in global scope shared among all players (the default scope is global)
```

File: README.md [hunk 10]
```diff
@@ -83,3 +84,90 @@ deathLinkService.OnDeathLinkReceived += (deathLinkObject) => {
...
+session.DataStorage["Add"] += 50; //Add 50 to the current value of `Add`
+session.DataStorage["Divide"] /= 2; //Divide current value of `Divide` in half
+session.DataStorage["Max"] <<= 80; //Set `Max` to 80 if the stored value is lower than 80
+session.DataStorage["Dictionary"] = JObject.FromObject(new Dictionary<string, int>()); //Set `Dictionary` to a Dictionary
+session.DataStorage["SetObject"] = JObject.FromObject(new SomeClassOrStruct()); //Set `SetObject` to a custom object
+session.DataStorage["BitShiftLeft"] += Bitwise.LeftShift(1); //Divide current value of `BitShiftLeft` in half
```
Comment (Ijwu) [inline README.md:140]:
```suggestion
session.DataStorage["BitShiftLeft"] += Bitwise.LeftShift(1); //Bitshift current value of  `BitShiftLeft` to left by 1
```

File: README.md [hunk 11]
```diff
@@ -83,3 +84,90 @@ deathLinkService.OnDeathLinkReceived += (deathLinkObject) => {
...
+session.DataStorage["Xor"] += Bitwise.Xor(0xFF); //Moddify `Xor` using the Bitwise excluse or operation
```
Comment (Ijwu) [inline README.md:141]:
```suggestion
session.DataStorage["Xor"] += Bitwise.Xor(0xFF); //Modify `Xor` using the Bitwise exclusive or operation
```

File: README.md [hunk 12]
```diff
@@ -83,3 +84,90 @@ deathLinkService.OnDeathLinkReceived += (deathLinkObject) => {
...
+* `= session.DataStorage["Key"]`, read value from the data storage
+
+Mathematical operations are supported using the following operators:
+* `+`, Add right value to left value
+* `-`, Subtract right value from left value
+* `*`, Multiply right value by left value
+* `/`, Divide right value by left value
+* `%`, Gets remainder after dividing left value by right value
+* `^`, Multiply left value by the power of the right value
+* `>>`, Override left with right value. if right value is lower
```
Comment (Ijwu) [inline README.md:109]:
```suggestion
* `>>`, Override left with right value, if right value is lower
```

File: README.md [hunk 13]
```diff
@@ -83,3 +84,90 @@ deathLinkService.OnDeathLinkReceived += (deathLinkObject) => {
...
+* `<<`, Override left with right value. if right value is bigger
```
Comment (Ijwu) [inline README.md:110]:
```suggestion
* `<<`, Override left with right value, if right value is bigger
```

File: README.md [hunk 14]
```diff
@@ -83,3 +84,90 @@ deathLinkService.OnDeathLinkReceived += (deathLinkObject) => {
...
+session.DataStorage["DifferentKey"] = session.DataStorage["A"] - 30; //Get value of `A`, Assign it to `DifferentKey` and then subtract 30
+session.DataStorage["Array"] = new []{ "One", "Two" }; //Arrays can be stored directly, List's needs to be converted ToArray() first 
+session.DataStorage["Array"] += new []{ "Three" }; //Append array values to existing array on the server
+
+//Chaining operations
+session.DataStorage["Min"] = (session.DataStorage["Min"] + 40) >> 100; //Add 40 to `Min`, then Set `Min` to 100 if `Min` is bigger than 100
+session.DataStorage["C"] = ((session.DataStorage["C"] - 6) + Bitwise.RightShift(1)) ^ 3; //Subtract 6 from `C`, then multiply `C` by 2 using bitshifting, then take `C` to the power of 3
+
+//Update callbacks
+//Enerylink deplete pattern, Subtract 50, then set value to 0 if its lower than 0
```
Comment (Ijwu) [inline README.md:151]:
```suggestion
//EnergyLink deplete pattern, subtract 50, then set value to 0 if its lower than 0
```

File: README.md [hunk 15]
```diff
@@ -83,3 +84,90 @@ deathLinkService.OnDeathLinkReceived += (deathLinkObject) => {
...
+	var changed = (int)new - (int)old; //Keep track of changes made to `OnChangeHandler` by any client, and calculate the difference
+};
+
+//Retrieving
+session.DataStorage.GetAsync<string>("Async", s => { string r = s }); //Retrieve value of `Async` asynchronously
+float f = session.DataStorage["Float"]; //Retrieve value for `Float` synchronously and store it as a float
+var d = session.DataStorage["DateTime"].To<DateTime>() //Retrieve value for `DateTime` as a DateTime struct
+var array = session.DataStorage["Strings"].To<string[]>() //Retrieve value for `Strings` as string Array
+
+//Handleing anonymous object, if the target type is not known you can use `To<JObject>()` and use its interface to access the members
```
Comment (Ijwu) [inline README.md:167]:
```suggestion
//Handling anonymous object, if the target type is not known you can use `To<JObject>()` and use its interface to access the members
```

**Review by Ijwu — CHANGES_REQUESTED** [review]
Sorry to nitpick, I had nothing else to complain about. This looks good, code wise! Wow, this is well done.

**Review by Ijwu — APPROVED** [review]
I am not willing to merge this in until 3.0 is merged in, but this is ready to go until then.

**Comment by JarnoWesthof** [comment]
Not sure why you see a dependency with 3.0 but ok

**Pull request merged**